Feature: Allow copying the table definition from the dropdown menu

nickterhaar opened the pull request:
## I have read the [CONTRIBUTING.md](https://github.com/supabase/supabase/blob/master/CONTRIBUTING.md) file.
YES

## What kind of change does this PR introduce?
Feature

## What is the current behavior?
https://github.com/supabase/supabase/issues/27538

We can now get the table definition from here
Would be great if we could copy it from here ("Copy table definition")

## What is the new behavior?
Now in the 'Table Editor', when opening the menu of one of the tables, you'll get the option to 'Copy Definition' (see screenshot).

## Additional context
I've used the same 'copy' icon as which is being used for 'Duplicate Table'. Because of this, I've put the 'Copy Definition' at the bottom of the list. Let me know if I need to use a different icon for this.

I've also created a 'useTableDefinition' hook, the logic of this hook was found in the 'TableDefinition.tsx' component. Because I had to reuse this logic, I thought it would be a better idea to write one new hook and let the existing component and my new feature make use of it, instead of reusing/copying the same code. Please let me know is this is okay or if I need to do it differently.

Add any other context or screenshots.
<img width="287" alt="copy-defintion" src="https://github.com/supabase/supabase/assets/2932357/93f21c8c-c5d0-4a32-b4a4-9f5573e52bba">

nickterhaar committed:
Adds button to TableEditorLayout -> EntityListItem to copy table definiton (functionality unfinished yet)Initial commit

nickterhaar committed:
Merge branch 'master' into 27538-Allow-copying-the-table-definition-from-the-dropdown-menu

nickterhaar committed:
Adds clipboard functionality to new Copy Definition button

nickterhaar committed:
Merge branch 'master' into 27538-Allow-copying-the-table-definition-from-the-dropdown-menu

nickterhaar committed:
Adds feature to allow copying the table definition from the dropdown menu

nickterhaar committed:
Feature: Allow copying the table definition from the dropdown menu

nickterhaar committed:
Merge branch 'master' into feature/Allow-copying-the-table-definition-from-the-dropdown-menu

saltcod commented on apps/studio/components/layouts/TableEditorLayout/EntityListItem.tsx:
```diff
@@ -314,6 +326,17 @@ const EntityListItem: ItemRenderer<Entity, EntityListItemProps> = ({
+                copyDefinition()
+              }}
+            >
+              <IconCopy size="tiny" />
```
I think let's use https://lucide.dev/icons/files instead?

nickterhaar replied:
I think that icon is usable. I'll change it.

nickterhaar replied:
Changed the icon to `IconClipboard`.
I've tried the files one from lucide, but it wasn't displaying in the correct color as the other icons. Also, because the function copy's the definition to the clipboard, I thought maybe this one was more suiting. Let me know.

saltcod commented on apps/studio/components/layouts/TableEditorLayout/EntityListItem.tsx:
```diff
@@ -29,6 +29,8 @@ import {
 import { useProjectContext } from '../ProjectLayout/ProjectContext'
 import { useProjectLintsQuery } from 'data/lint/lint-query'
 import { getEntityLintDetails } from 'components/interfaces/TableGridEditor/TableEntity.utils'
+import useTableDefinition from '../../../hooks/misc/useTableDefinition'
```
should be able to import these like this?

https://github.com/supabase/supabase/blob/master/apps/studio/components/interfaces/TableGridEditor/GridHeaderActions.tsx#L22

nickterhaar replied:
I'll change the import as it is in the given example.
Got mine from somewhere else where it's also done the way I did it, but how it's done in the example is much cleaner.

saltcod commented:
Thanks @nickterhaar !   Left a couple of comments

nickterhaar commented on apps/studio/hooks/misc/useTableDefinition.ts:
```diff
@@ -0,0 +1,66 @@
+import { useMemo } from 'react'
```
The `useMemo` is what was used in the logic I got from the 'TableDefinition.tsx'. But I'll take a look into using the 'table-definition-query.ts'

nickterhaar replied:
Removed the use of `useMemo`. We still need the rest to have a `tableResult` or a `viewResult`. Also the `useTableDefinitionQuery` returns a not so nice formatted (unnecessary spaces, etc.) , uppercase version of the query, so `formatDefinition` is still used to make it look as nice as it already was in the definition tab.

nickterhaar commented:
> Thanks @nickterhaar ! Left a couple of comments

@saltcod Thanks for the comments. Gonna work on them asap.
Would it also be a welcoming addition to the Database where there is also a dropdown with almost the same option as in the Table Editor? Let me know if you want me to also add this option to that menu.

nickterhaar committed:
Merge branch 'master' into feature/Allow-copying-the-table-definition-from-the-dropdown-menu

nickterhaar committed:
Icon changed to IconClipboard

nickterhaar committed:
Cleaned up hook imports

nickterhaar committed:
Removed the use of 'useMemo' from 'useTableDefinition'

saltcod commented:
yeah sure @nickterhaar making the dropdowns consistent would be great. 
Feel free to do in a separate PR though — we can get this one in now and avoid any additional conflicts

nickterhaar commented:
> yeah sure @nickterhaar making the dropdowns consistent would be great. Feel free to do in a separate PR though — we can get this one in now and avoid any additional conflicts

Alright! I will start working on that soon and will create a new PR for that.

Is there anything else that I can do for this PR? If so, let me know and I will do that first.

nickterhaar committed:
Changed icon to lucide-react 'files' & merge with master

nickterhaar committed:
Clean up of how 'const formattedDefinition' is set

saltcod commented:
@nickterhaar  any idea what happened with all the file changes? There's 120 file changes in this PR now 😬

nickterhaar commented:
> @nickterhaar any idea what happened with all the file changes? There's 120 file changes in this PR now 😬

@saltcod I merged master with my branch and suddenly had 'some' extra files. My mistake.

saltcod commented:
@nickterhaar 
With all the conflicts and changes, I think you might be better recreating this branch, cherry picking the good commits?

nickterhaar committed:
Revert "Changed icon to lucide-react 'files' & merge with master"

This reverts commit ffc11384ddc99f3e630f2aff4bb94bff9fb059cf.

nickterhaar committed:
Changed icon to lucide-react 'files'

nickterhaar commented:
@saltcod I've reverted the merge with master commit, changed the icon in `EntityListItem.tsx` back to lucide 'files'. So the amount of changed files is back to 3 again and all 3 should be and work as intended.

nickterhaar commented:
See https://github.com/supabase/supabase/pull/27900 for the same functionality on the Database page.

Hallidayo commented:
Hi @nickterhaar - Are you able to look at this conflict?

nickterhaar commented:
@Hallidayo I'll have a look soon, probably this weekend.

nickterhaar commented:
@Hallidayo Sorry, I did my best but this conflict resolve is/was a mess to rebase/merge.

nickterhaar committed:
Fix

nickterhaar committed:
Fix

nickterhaar committed:
Fix

nickterhaar committed:
Fix

nickterhaar commented:
@Hallidayo I tried to resolve the conflicts, but everything that is mentioned [here](https://github.com/supabase/supabase/pull/27621/conflicts) as a conflict is exactly what is needed for this feature to work. Please help me with what I'm doing wrong or tell me what to look for.

Hallidayo commented:
Hi @nickterhaar - Is it worth maybe closing this branch/pull and then opening a new branch/pull then cherry pick over your changes?

nickterhaar commented:
@Hallidayo I'll have a look, but that will probably be next weekend. I'll keep you posted.

nickterhaar commented:
@Hallidayo I did a redo of this PR [here](https://github.com/supabase/supabase/pull/30393).
